form/create_custom_choice_type.rst

```diff
@@ -0,0 +1,273 @@
+    {
+        // This must be passed by the type
+        // this loader won't be registered as service
+        private $api;
+        // define more options if needed
+        private $someOption;
+
+        public function __construct(AcmeApi $api, string $someOption)
+        {
+            $this->api = $api;
+            $this->someOption = $someOption;
+        }
+
+        protected function loadChoices(): iterable
+        {
+            return $this->api->loadCategories($this->someOption));
+        }
+
+        protected function doLoadChoicesForValues(array $values): array
+        {
+            return $this->api->loadCategoriesForNames($values, $this->someOption);
+        }
+
+        protected function doLoadValuesForChoices(array $choices): array
+        {
+            $values = [];
+
+            // compute string values that must be submitted
+
+            return $values;
+        }
+    }
+
+Here we implement three protected methods:
+
+`loadChoices(): iterable`
+
+    This method is abstract and is the only one that needs to be implemented,
+    it is called when the list is fully loaded (i.e when rendering the view).
+    It must return an array or a traversable object, keys are default labels
+    unless the :ref:`choice_label <reference-form-choice-label>` option is
+    defined.
+    Choices can be grouped with keys as group name and nested iterable choices
+    in alternative to the :ref:`group_by <reference-form-group-by>` option.
+
+`doLoadChoicesForValues(array $values): array`
+
+    Optional, to improve performance this method is called when the data is
+    submitted. You can then load the choices partially, by using the submitted
+    values passed as only argument.
+    The list is fully loaded by default.
+
+`doLoadValuesForChoices(array $choices): array`
+
+    Optional, in alternative to the
+    `choice_value <reference-form-choice-value>` option.
+    You can implement this method to return the string values partially, the
+    initial choices are passed as only argument.
+    The list is fully loaded by default unless the ``choice_value`` option is
+    defined.
+
+Then you need to update the previous example to use the new loader instead::
+
+    // src/Form/Type/CategoryChoiceType.php;
+
+    // ... same as before
+    use App\Api\AcmeApi;
+    use App\Form\ChoiceList\AcmeCategoryLoader;
+
+    class CategoryChoiceType extends AbstractType
+    {
+        // using the default configuration, the type is a service
+        // so the api will be autowired
+        private $api;
+
+        public function __construct(AcmeApi $api)
+        {
+            $this->api = $api;
+        }
+
+        // ... same as before
+
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver
+                // ... same as before
+                // but use the custom loader instead
+                ->setDefault('choice_loader', static function(Options $options) {
```

Comment:
Do we really want `static function(...) {}` in the code? I know it's how it's done in the core code, but I think it adds complexity without benefit.
Docs are imho somewhere in the middle between not providing bad practices, but also not providing the perfect practice to keep things simple enough to be understand (and does static even work in this case, as we're using `$this` and `$this->api` in the closure body?)

Reply:
I've added them afterwards, and introduced that mistake, whereas I usually do this with caution... Well I wouldn't check and update all the docs, but I would be in favor to consider this a good practice and use it when relevant, e.g showing example userland code that may be copy/pasted like here and not using `$this`.